src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -1027,14 +1027,18 @@ public IQueryCache GetQueryCache(string cacheRegion)
 				return null;
 			}
 
+			// The factory may be run concurrently by threads trying to get the same region.
+			// But the GetOrAdd will yield the same lazy for all threads, so only one will
+			// initialize. https://stackoverflow.com/a/31637510/1178314
 			return queryCaches.GetOrAdd(
 				cacheRegion,
-				cr =>
-				{
-					IQueryCache currentQueryCache = settings.QueryCacheFactory.GetQueryCache(cacheRegion, updateTimestampsCache, settings, properties);
-					allCacheRegions[currentQueryCache.RegionName] = currentQueryCache.Cache;
-					return currentQueryCache;
-				});
+				cr => new Lazy<IQueryCache>(
+					() =>
+					{
+						var currentQueryCache = settings.QueryCacheFactory.GetQueryCache(cr, updateTimestampsCache, settings, properties);
+						allCacheRegions[currentQueryCache.RegionName] = currentQueryCache.Cache;
```

Comment:
I think this needs to go out of the lambda.

Reply:
The lazy lambda is the thing guaranteed to be executed only once. It looks to me as the natural place for putting access needing to be protected (here, guaranteeing we store the cache uniquely (only for all queries in this PR) built for the region).
Getting it out of the lambda would either mean re-storing it at each cache access, which looks inefficient, or will require some additional code like
```c#
// The factory may be run concurrently by threads trying to get the same region.
// But the GetOrAdd will yield the same lazy for all threads, so only one will
// initialize. https://stackoverflow.com/a/31637510/1178314
var lazyCache = queryCaches.GetOrAdd(
	cacheRegion,
	cr => new Lazy<IQueryCache>(
		() => settings.QueryCacheFactory.GetQueryCache(cr, updateTimestampsCache, settings, properties)));
if (lazyCache.IsValueCreated)
	return lazyCache.Value;
var cache = lazyCache.Value;
allCacheRegions[currentQueryCache.RegionName] = cache.Cache;
return cache;
```
This causes an overhead for all accesses to a cache by checking the `IsValueCreated` (maybe negligible, I do not know if this property is very lightweight or not), compared to the "fully initialized in lambda" one.
(And this one may also store multiple times the generated region cache, without creating an issue excepted slight inefficiency.)
Or using a boolean:
```c#
// The factory may be run concurrently by threads trying to get the same region.
// But the GetOrAdd will yield the same lazy for all threads, so only one will
// initialize. https://stackoverflow.com/a/31637510/1178314
var added = false;
var lazyCache = queryCaches.GetOrAdd(
	cacheRegion,
	cr => new Lazy<IQueryCache>(
		() => 
		{
			added = true;
			return settings.QueryCacheFactory.GetQueryCache(cr, updateTimestampsCache, settings, properties);
		}));
var cache = lazyCache.Value;
if (added)
	allCacheRegions[currentQueryCache.RegionName] = cache.Cache;
return cache;
```
So now the overhead is very small, just a boolean check. But there we capture a local variable into the lambda for mutating it, which is far more debatable than just doing the whole initialization in it.

Moreover, both propositions above are highly sensitive to code lines order: the `IsValueCreated` must of course be checked before accessing `Value`, while the `added` boolean must be checked after.

So those other way of implementing it look to me as only having disadvantages.

Maybe would you just want this lambda to be converted as full blown function?

```c#
	// The factory may be run concurrently by threads trying to get the same region.
	// But the GetOrAdd will yield the same lazy for all threads, so only one will
	// initialize. https://stackoverflow.com/a/31637510/1178314
	return queryCaches.GetOrAdd(
		cacheRegion,
		cr => new Lazy<IQueryCache>(
			() => CreateQueryCache(cr))).Value;
}

private IQueryCache CreateQueryCache(string region)
{
	var cache = settings.QueryCacheFactory.GetQueryCache(cr, updateTimestampsCache, settings, properties);
	allCacheRegions[currentQueryCache.RegionName] = cache.Cache;
	return cache;
}
```

Or something else? But I do not see what.